### PR TITLE
Added unit tests for NavigationView.

### PR DESCRIPTION
**Pull request by cafesilencio**
### Description
Added some unit tests for `NavigationView`.

### CODE REVIEW

File: libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/NavigationViewTest.kt [hunk 1]
```diff
@@ -1,62 +1,297 @@
...
+            "token",
+            MapInitOptions(ctx),
+            NavigationViewOptions.Builder().build()
+        )
+        navigationView.addTripSessionStateObserver(observer)
+        assertEquals(1, navigationView.externalTripSessionStateObservers.size)
+
+        navigationView.removeTripSessionStateObserver(observer)
+
+        assertEquals(0, navigationView.externalTripSessionStateObservers.size)
```
Comment (VysotskiVadim) [inline libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/NavigationViewTest.kt:295]: I guess this test won't help you with refactoring. If you change the way you store observers - tests will just break. But the whole point of testing is to help you refactor. Do you want change implementation if you know that you have to rewrite a bunch of tests and retest everything manually? 

Maybe it's better to check if callback was called or was not called when something happened?

Comment (cafesilencio) [reply libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/NavigationViewTest.kt:211]: I'm testing the method does what it's supposed to do. There is no callback. If something in the implementation changes I want the test to break.

Comment (VysotskiVadim) [reply libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/NavigationViewTest.kt:211]: sorry, I meant to check if the observer is called, not a callback.

Comment (VysotskiVadim) [reply libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/NavigationViewTest.kt:211]: > I'm testing the method does what it's supposed to do.

There are could be a few answers to the question "What are the methods `addRouteProgressObserver` `removeTripSessionStateObserver` supposed to do?".

You as an author think that those methods should add and remove observers in the array. That's a fair point. It's how methods are implemented right now.

What if you ask our users? I doubt that answer will be the same. I guess users want a route progress observer to be called when the route progress changes. I don't think they care about our internal implementation.

Why don't we test behavior that users expect from the methods? Tests shouldn't let us change the behavior which our users expect. Tests should let us change an internal implementation that users don't care about. That's actually a definition of refactoring: changes of implementation that don't affect behavior.

When you write tests that check implementation details you:
1. Don't test behavior that users expect.
2. Don't let us refactor later. Changes to implementations will make the test red, despite the fact the feature still works as users expect.

Comment (LukasPaczos) [reply libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/NavigationViewTest.kt:209]: I'd agree with what @VysotskiVadim mentioned above. The clear-cut argument for me is the need to add `@VisibleForTesting`. I'd prefer us limiting then usage of this annotation (and similar mechanisms) as much as possible. We shouldn't testing that we have subscribed to some type of an internal list. We should rather try to generate the relevant event (from a mocked `MapboxNavigaion` in this case) and make sure that it's delivered to the developer.

In short, I think that we should capture and invoke the observer registered with `MapboxNavigation` and verify that the observer that user registered has been invoked. Regardless of what was the intermediate step (in this case registering to `externalVoiceInstructionsObservers`).

**Comment by cafesilencio** [comment]
I haven't been able to get the tests to pass in CI for some reason. In addition this doesn't look like a component we'll be able to successfully test with unit tests so we'll have to develop instrumentation tests instead. Closing this.
